backend/cmm_helpers.ml

```diff
@@ -4549,31 +4558,40 @@ let store ~dbg kind init ~addr ~new_value =
   Cop (Cstore (kind, init), [addr; new_value], dbg)
 
 let direct_call ~dbg ty pos f_code_sym args =
-  Cop (Capply (ty, pos), f_code_sym :: args, dbg)
+  Cop
+    ( Capply { result_type = ty; region = pos; callees = Some [f_code_sym] },
+      Cconst_symbol (f_code_sym, dbg) :: args,
+      dbg )
 
 let indirect_call ~dbg ty pos alloc_mode f args_type args =
   might_split_call_caml_apply ty args_type Asttypes.Mutable f args pos
     alloc_mode dbg
 
-let indirect_full_call ~dbg ty pos alloc_mode f args_type args =
-  match args_type with
-  (* the single-argument case is already optimized by indirect_call *)
-  | [_] -> indirect_call ~dbg ty pos alloc_mode f args_type args
-  | [] -> Misc.fatal_error "indirect_full_call: args_type was empty"
-  | _ :: _ :: _ ->
-    (* Use a variable to avoid duplicating the cmm code of the closure [f]. *)
-    let v = Backend_var.create_local "*closure*" in
-    let v' = Backend_var.With_provenance.create v in
-    (* get the function's code pointer *)
-    let fun_ptr =
-      load ~dbg Word_int Asttypes.Mutable ~addr:(field_address (Cvar v) 2 dbg)
+let indirect_full_call ~dbg ty pos f ~callees args_type args =
+  (* Use a variable to avoid duplicating the cmm code of the closure [f]. *)
+  let v = Backend_var.create_local "*closure*" in
+  let v' = Backend_var.With_provenance.create v in
+  (* get the function's code pointer *)
+  let fun_ptr =
+    let offset =
+      match args_type with
+      | [_] -> 0
```

Comment:
I've noticed in passing that the new version is no longer calling `indirect_call`.  Is this intentional?